431203: CVE-2007-6698 openldap: slapd crash on NOOP control operation on
entry in bdb storage

Source: Ralf Haferkamp <rhafer@suse.de>
Reviewed-By: Jan Safranek <jsafrane@redhat.com>

Index: openldap-2.2.24/servers/slapd/back-bdb/modify.c
===================================================================
--- openldap-2.2.24.orig/servers/slapd/back-bdb/modify.c
+++ openldap-2.2.24/servers/slapd/back-bdb/modify.c
@@ -562,6 +562,08 @@ retry:	/* transaction retry */
 			rs->sr_text = "txn_abort (no-op) failed";
 		} else {
 			rs->sr_err = LDAP_NO_OPERATION;
+			rs->sr_text = NULL;
+			ltid = NULL;
 			goto return_results;
 		}
 	} else {
@@ -660,8 +664,8 @@ done:
 			ch_free( pm_prev );
 		}
 		TXN_ABORT( ltid );
-		op->o_private = NULL;
 	}
+	op->o_private = NULL;
 
 	if( e != NULL ) {
 		bdb_unlocked_cache_return_entry_w (&bdb->bi_cache, e);
Index: openldap-2.2.24/servers/slapd/back-bdb/ctxcsn.c
===================================================================
--- openldap-2.2.24.orig/servers/slapd/back-bdb/ctxcsn.c
+++ openldap-2.2.24/servers/slapd/back-bdb/ctxcsn.c
@@ -52,6 +52,10 @@ bdb_csn_commit(
 
 	assert( !BER_BVISNULL( &op->o_bd->be_context_csn ) );
 
+	if ( op->o_noop ) {
+	    return BDB_CSN_COMMIT;
+	}
+
 	if ( ei ) {
 		e = ei->bei_e;
 	}
Index: openldap-2.2.24/servers/slapd/back-bdb/add.c
===================================================================
--- openldap-2.2.24.orig/servers/slapd/back-bdb/add.c
+++ openldap-2.2.24/servers/slapd/back-bdb/add.c
@@ -400,6 +400,8 @@ retry:	/* transaction retry */
 			rs->sr_text = "txn_abort (no-op) failed";
 		} else {
 			rs->sr_err = LDAP_NO_OPERATION;
+			rs->sr_text = NULL;
+			ltid = NULL;
 			goto return_results;
 		}
 
@@ -481,8 +483,8 @@ return_results:
 done:
 	if( ltid != NULL ) {
 		TXN_ABORT( ltid );
-		op->o_private = NULL;
 	}
+	op->o_private = NULL;
 
 	if( postread_ctrl != NULL ) {
 		slap_sl_free( (*postread_ctrl)->ldctl_value.bv_val, op->o_tmpmemctx );
Index: openldap-2.2.24/servers/slapd/back-bdb/delete.c
===================================================================
--- openldap-2.2.24.orig/servers/slapd/back-bdb/delete.c
+++ openldap-2.2.24/servers/slapd/back-bdb/delete.c
@@ -485,6 +485,8 @@ retry:	/* transaction retry */
 			rs->sr_text = "txn_abort (no-op) failed";
 		} else {
 			rs->sr_err = LDAP_NO_OPERATION;
+			rs->sr_text = NULL;
+			ltid = NULL;
 			goto return_results;
 		}
 	} else {
@@ -584,8 +586,8 @@ done:
 
 	if( ltid != NULL ) {
 		TXN_ABORT( ltid );
-		op->o_private = NULL;
 	}
+	op->o_private = NULL;
 
 	slap_sl_free( ctxcsn_ndn.bv_val, op->o_tmpmemctx );
 
Index: openldap-2.2.24/servers/slapd/back-bdb/modrdn.c
===================================================================
--- openldap-2.2.24.orig/servers/slapd/back-bdb/modrdn.c
+++ openldap-2.2.24/servers/slapd/back-bdb/modrdn.c
@@ -833,6 +833,8 @@ retry:	/* transaction retry */
 			rs->sr_text = "txn_abort (no-op) failed";
 		} else {
 			rs->sr_err = LDAP_SUCCESS;
+			rs->sr_text = NULL;
+			ltid = NULL;
 			goto return_results;
 		}
 
@@ -981,8 +985,8 @@ done:
 			ch_free( pm_prev );
 		}
 		TXN_ABORT( ltid );
-		op->o_private = NULL;
 	}
+	op->o_private = NULL;
 
 	if( preread_ctrl != NULL ) {
 		slap_sl_free( (*preread_ctrl)->ldctl_value.bv_val, op->o_tmpmemctx );
